### PR TITLE
[FEAT] Expander - Parameter expansion and Quote removal

### PR DESCRIPTION
**Pull request by itislu**
#### The expander handles the following cases:
| Input | Expanded |
|--------|--------|
| `$USER` | `<username>` |
| `${USER}` | `<username>` |
| `$?` | `<exit code>` |
| `$` | `$` |
| `$.USER` | `$.USER` |
| `$abc` | `NULL` | 
| `$abc$USER` | `<username>` |
| `$"USER"` | `USER` |
| `$'USER'` | `USER` |
| `${USER?abc}` | `<username>` |
| `${??abc}` | `<exit code>` | 
| `${??$USER}` | `<exit code>` |
| `"$USER"` | `<username>` |
| `'$USER'` | `$USER` |
| `$"$USER"` | `<username>` |
| `$'$USER'` | `$USER` | 
| `"'$USER'"` | `'<username>'` |
| `'$"$USER"'` | `$"$USER"` |
| `"$'$USER'"` | `$'<username>'` |
| `$"{USER}"` | `{USER}` |
| `$'{USER}'` | `{USER}` |
| `"$'{USER}'"` | `$'{USER}'` |
| `"$'${USER}'"` | `$'<username>'` |
| `"$'{$USER}'"` | `$'{<username>}'` |

#### These cases cause a "`%s`: bad substitution" error:
| Input | `%s` |
|--------|--------|
| `${USER.}` | `${USER.}` |
| `${USER.}abc` | `${USER.}abc` |
| `"${USER.}"abc` | `${USER.}` |
| `"$'$USER${USER.}'"` | `$'$USER${USER.}'` |
| `$"'$USER${USER.}'"` | `'$USER${USER.}'` |
| `${?USER}` | `${?USER}` |
| `"$'${$USER}'"` | `$'${$USER}'` |

#### Quote removal:
| Input | Expanded |
|--------|--------|
| `""` | `<empty string>` |
| `''` | `<empty string>` |
| `"''"` | `''` |
| `'""'` | `""` |

#### There are still open questions to resolve:
- [ ] `echo ${abc.$USER`
 		At the moment it will just leave the string as is. Should probably already stop at lexer with syntax error.
- [x] Where to handle the "%s: bad substitution" error message, e.g. `${abc.}`?
- [ ]  This test case is terrible, bc it changes a cmd arg to a cmd name:
  ```
  bash-5.1$ $abc 123
  bash: 123: command not found
  ```

### CODE REVIEW

File: source/utils/string_utils.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		while (str[*i])
+		{
+			(*i)++;
+			if (str[*i] == *open_quote)
+			{
+				(*i)++;
+				break ;
+			}
+		}
+	}
```
Comment (itislu) [inline source/utils/string_utils.c:41]: Need to reduce indentation.

File: source/debug/print_expanded_cmd_table.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	while (node)
+	{
+		cmd_table = (t_cmd_table *)node->content;
+		// if (cmd_table->type == C_SIMPLE_CMD)
+		// {
+			printf("\033[1;31mBEFORE EXPANSION:\n\033[0m");
+			printf("\033[0;31m========= %d =========\n\033[0m", cmd_table->id);
+			print_cmd_table(cmd_table);
+			printf("\033[0;31m=====================\n\033[0m");
+			if (!expand_cmd_table(cmd_table, shell))
```
Comment (itislu) [inline source/debug/print_expanded_cmd_table.c:98]: Should not modify the command table

**Comment by LeaYeh** [comment]
Known issue that will effect the interface, need to be solved

```sh
bash-5.1$ ${} && echo 1
bash: ${}: bad substitution
bash-5.1$ ${} || echo 1
bash: ${}: bad substitution
bash-5.1$ ${} | echo 1
1
bash: ${}: bad substitution
```

**Review by LeaYeh — CHANGES_REQUESTED** [review]
Known issues that will effect the interface, need to be solved

```sh
bash-5.1$ ${} && echo 1
bash: ${}: bad substitution
bash-5.1$ ${} || echo 1
bash: ${}: bad substitution
bash-5.1$ ${} | echo 1
1
bash: ${}: bad substitution
```

```sh
bash-5.1$ $aaaaa || echo 1
bash-5.1$ $aaaaa && echo 1
1
```

```sh
bash-5.1$ "$abc"
bash: : command not found
```

Before we figure out these issues we can't decide the interface.

**Comment by LeaYeh** [comment]
Maybe the expander could return...
1. list == NULL
    * executor should execute `true` as command, means do nothing and exit==0
2. list->content == NULL
    *  executor will use the first non-null content as command name and others are command args
    * empty string has different meaning with NULL

**Comment by itislu** [comment]
Expansion happens during the executor. 
Proof:
```sh
bash-5.1$ echo 1 && echo ${a.b} || echo 1
1
bash: ${a.b}: bad substitution
```

**Comment by itislu** [comment]
Executor needs to filter out all `node->content == NULL` nodes and remove them from the list.
```sh
bash-5.1$ echo $aaa 1
1
```

**Comment by itislu** [comment]
How NULL should behave for redirections:
```sh
bash-5.1$ echo 1 > abc > $aaa > def
bash: $aaa: ambiguous redirect
bash-5.1$ ls
abc
bash-5.1$ cat abc
```
```sh
bash-5.1$ echo 1 > abc > $USER > def
bash-5.1$ ls
abc  def  ldulling
bash-5.1$ cat def
1
```

**Comment by itislu** [comment]
#### Solution for the `bad substitution` problem:

If the expander encounters a bad substitution, it should return `false`, but also set the string it got to `NULL`.
The executor then can check the following:
- If the expander returns `false` and the string got turned to `NULL`, an expander error like `bad substitution` happened.
  The executor should do nothing else for the current command table anymore, and go to the next command table.

- Otherwise, if the expander returns `false` and the string got NOT turned to `NULL`, a critical issue occured, like a malloc fail, and the shell needs to exit.

File: include/expander.h [hunk 3]
```diff
@@ -15,16 +15,27 @@
 
 # include "defines.h"
 
+# define GET	0
+# define TOGGLE	1
+# define RESET	2
+
```
Comment (itislu) [inline include/expander.h:21]: @LeaYeh What do you think, are these defines ok here?

File: source/debug/print_expanded_cmd_table.c (hunk 2 above)
Comment (itislu) [reply source/debug/print_expanded_cmd_table.c:98]: @LeaYeh Please help me where we should put the expander test function for the moment.

Comment (LeaYeh) [reply source/debug/print_expanded_cmd_table.c:98]: Modify to the static function

**Comment by LeaYeh** [comment]
The interface of expander...
- if true...
  1. string=string, *list = one element list (except wildcard)
- if false...
  1. general error, string=NULL, *list = NULL
  2. expander error, string=string, *list = NULL

TODO list for executor...
1. after expander if cmd_name => NULL, the cmd_name should be replace by the first non-null argument
2. cmd_args need to filter out the NULL
=> Maybe combine the cmd_name and cmd_args into one list in a function and pass to expander 1-1 then separate back
3. if the executor receives "1 2 3" as cmd_name, it should be split into "1" "2" "3"

File: include/expander.h (hunk 3 above)
Comment (LeaYeh) [reply include/expander.h:21]: use enum

**Comment by LeaYeh** [comment]
Need to be checked...
Should we handle the different expansions by the scenario in expander or just in executor?
* in heredoc, the quotes will be removed but not expand

**Comment by itislu** [comment]
- [ ] Need to provide the submodules of the expander as separate interfaces for the executor.

File: source/frontend/expander/expander_utils.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		while (is_valid_varname(str[replace_len]))
+			replace_len++;
+	return (replace_len);
+}
+
+void	free_and_reset(char *dup, char **str)
+{
+	free(dup);
+	ft_free_and_null((void **)str);
+	is_open_pair('"', RESET);
```
Comment (LeaYeh) [inline source/frontend/expander/expander_utils.c:56]: Maybe reset the whole sets may be more clear

**Review by LeaYeh — CHANGES_REQUESTED** [review]

**Review by LeaYeh — APPROVED** [review]

**Pull request merged**